src/models/SkipGANomaly.jl

```diff
@@ -0,0 +1,315 @@
+        loss_for_real + loss_for_fake + lat_loss, con_loss, lat_loss
+end
+
+"""
+    function anomaly_score(SkipGAN::SkipGANomaly, real_input; lambda = 0.5)
+computes unscaled anomaly score of real input, losses are weighted by factor lambda
+"""
+function anomaly_score(SkipGAN::SkipGANomaly, real_input; lambda = 0.9, dims=[1,2,3])
+    fake = SkipGAN.generator(real_input) # + randn(typeof(real_input[1]), size(real_input))
+
+    pred_real, feat_real = SkipGAN.discriminator(real_input)
+    pred_fake, feat_fake = SkipGAN.discriminator(fake)
+
+	rec_loss  = Flux.mae(fake, real_input, agg=x->Flux.mean(x, dims=dims)) # reconstruction loss -> similar to contextual loss
+    lat_loss = Flux.mse(feat_fake, feat_real, agg=x->Flux.mean(x, dims=dims)) # loss of latent representation
+	return lambda .* rec_loss .+ (1 - lambda) .* lat_loss |>cpu
+end
+
+function generalized_anomaly_score(SkipGAN::SkipGANomaly, real_input; R="mae", L="mae", lambda=0.9, dims=[1,2,3])
+	fake = SkipGAN.generator(real_input) # + randn(typeof(real_input[1]), size(real_input))
+
+    pred_real, feat_real = SkipGAN.discriminator(real_input)
+    pred_fake, feat_fake = SkipGAN.discriminator(fake)
+
+	R = getfield(Flux, Symbol(R))
+ 	L = getfield(Flux, Symbol(L))
+	rec_loss  = R(fake, real_input, agg=x->Flux.mean(x, dims=dims)) # reconstruction loss -> similar to contextual loss
+    lat_loss = L(feat_fake, feat_real, agg=x->Flux.mean(x, dims=dims)) # loss of latent representation
+	return lambda .* rec_loss .+ (1 - lambda) .* lat_loss |>cpu
+end
+"""
+	Model's training and inference
+"""
+
+"""
+	fit!(SkipGAN::SkipGANomaly, data, params)
+"""
+function StatsBase.fit!(SkipGAN::SkipGANomaly, data, params)
+    # prepare batches & loaders
+    train_loader, val_loader = prepare_dataloaders(data, params)
+    # training info logger
+    history = GANomalyHistory()
+    # prepare for early stopping
+    best_model = deepcopy(SkipGAN)
+    patience = params.patience
+    best_val_loss = 1e10
+	val_batches = length(val_loader)
+    # define optimiser
+	opt = Flux.Optimise.ADAM(params.lr)
+
+	ps_g = Flux.params(SkipGAN.generator)
+	ps_d = Flux.params(SkipGAN.discriminator)
+
+	progress = Progress(length(train_loader))
+	for (iters, X) in enumerate(train_loader)
+		#generator update
+		loss1, back = Flux.pullback(ps_g) do
+			generator_loss(SkipGAN, getobs(X)|>gpu, weights=params.weights)
+		end
+		grad = back((1f0, 0f0, 0f0, 0f0))
+		Flux.Optimise.update!(opt, ps_g, grad)
+
+		# discriminator update
+		loss2, back = Flux.pullback(ps_d) do
+			discriminator_loss(SkipGAN, getobs(X)|>gpu)
+		end
+		grad = back(1f0)
+		Flux.Optimise.update!(opt, ps_d, grad)
+
+		history = update_history(history, loss1, loss2)
+		next!(progress; showvalues=[(:iters, "$(iters)/$(params.iters)"),
+									(:generator_loss, loss1[1]),
+									(:discriminator_loss, loss2)
+									])
+		#TODO optionaly add discriminator restrart if its loss drops under 1e-5
+		if mod(iters, params.check_every) == 0
```

Comment:
Inconsistent indentation, a bit hard to read. How does the early stopping work here?

Reply:
Here it is bit tricky(resp messy) but while i checking checking early stopping criterion `anomality = (params.lambda*tot_val_loss_rl + (1-params.lambda)*tot_val_loss_ll)` 
i am also computing generator and discriminator loss on validation dataset with normal observations.
I thought it can be useful to trace it as well.

Reply:
I see, this makes sense.